File: app/src/main/java/com/omar/retromp3recorder/app/main/MainActivity.java
package com.omar.retromp3recorder.app.main;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.media.audiofx.Visualizer;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v4.util.Pair;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.ScrollView;
import android.widget.TextView;

import com.github.alkurop.jpermissionmanager.PermissionOptionalDetails;
import com.github.alkurop.jpermissionmanager.PermissionRequiredDetails;
import com.github.alkurop.jpermissionmanager.PermissionsManager;
import com.jakewharton.rxbinding3.view.RxView;
import com.omar.retromp3recorder.app.App;
import com.omar.retromp3recorder.app.R;
import com.omar.retromp3recorder.app.customviews.VisualizerView;
import com.omar.retromp3recorder.app.recording.recorder.VoiceRecorder;
import com.omar.retromp3recorder.app.utils.NotUnitTestable;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.annotations.Nullable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

import static com.omar.retromp3recorder.app.utils.VarargHelper.createHashMap;
import static io.reactivex.android.schedulers.AndroidSchedulers.mainThread;
import static io.reactivex.schedulers.Schedulers.computation;

public class MainActivity extends AppCompatActivity implements MainView {
    private static final int DELAY_MILLIS = 150;
    private final PublishSubject<Action> actionPublishSubject = PublishSubject.create();
    private final Subject<State> stateSubject = BehaviorSubject.create();
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    private final Handler scrollDownHandler = new Handler();

    private ImageView playButton, recordButton, shareButton;
    private LinearLayout logContainerView, sampleRateContainer, bitRateContainer;
    private List<RadioButton> sampleRateGroup;
    private List<RadioButton> bitRateGroup;
    private ScrollView scrollView;
    private VisualizerView visualizerView;
    private ImageView background;

    @Nullable
    private Visualizer visualizer;
    private PermissionsManager permissionsManager;
    private HashMap<String, PermissionOptionalDetails> permissionsMap;
    private LayoutInflater layoutInflater;

    @Inject
    MainViewInteractor interactor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        layoutInflater = LayoutInflater.from(this);
        App.appComponent.inject(this);
        permissionsManager = new PermissionsManager(this);
        findViews();
        setListeners();
        createPermissionsMap();
        Picasso.with(this).load(R.drawable.bg).fit().into(background);
        Disposable disposable = actionPublishSubject
                .compose(interactor.process())
                .compose(MainViewResultMapper.map())
                .observeOn(mainThread())
                .subscribe(this::renderView);
        compositeDisposable.add(disposable);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        permissionsManager.onRequestPermissionsResult(requestCode, permissions, grantResults);
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @android.support.annotation.Nullable Intent data) {
        permissionsManager.onActivityResult(requestCode);
        super.onActivityResult(requestCode, resultCode, data);
    }

    private void createPermissionsMap() {
        permissionsMap = createHashMap(
                new Pair<>(Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        new PermissionRequiredDetails(
                                getString(R.string.write_permission_title),
                                getString(R.string.write_permission_message),
                                getString(R.string.write_required_message))
                ),
                new Pair<>(Manifest.permission.RECORD_AUDIO,
                        new PermissionRequiredDetails(
                                getString(R.string.record_permission_title),
                                getString(R.string.record_permission_message),
                                getString(R.string.record_required_message)
                        )
                )
        );
    }

    private void findViews() {
        playButton = findViewById(R.id.iv_play);
        recordButton = findViewById(R.id.iv_record);
        shareButton = findViewById(R.id.iv_share);
        logContainerView = findViewById(R.id.ll_logHolder);
        scrollView = findViewById(R.id.scrollView);
        visualizerView = findViewById(R.id.visualizer);
        sampleRateContainer = findViewById(R.id.left_radio_container);
        bitRateContainer = findViewById(R.id.right_radio_container);
        background = findViewById(R.id.background);
        prefillRadioButtons();
    }

    @SuppressLint("DefaultLocale")
    private void prefillRadioButtons() {
        addTitleView(bitRateContainer, getString(R.string.bit_rate));
        addTitleView(sampleRateContainer, getString(R.string.sample_rate));

        sampleRateGroup = new ArrayList<>();
        bitRateGroup = new ArrayList<>();

        for (VoiceRecorder.SampleRate sampleRate : VoiceRecorder.SampleRate.values()) {
            String title = getString(
                    R.string.sample_rate_format,
                    sampleRate.value
            );
            RadioButton radioButton = addCheckBox(sampleRateContainer, title);
            sampleRateGroup.add(radioButton);
        }

        for (VoiceRecorder.BitRate bitRate : VoiceRecorder.BitRate.values()) {
            String title = getString(
                    R.string.bit_rate_format,
                    bitRate.value
            );
            RadioButton radioButton = addCheckBox(bitRateContainer, title);
            bitRateGroup.add(radioButton);
        }
    }

    private RadioButton addCheckBox(ViewGroup container, String text) {
        @SuppressLint("InflateParams")
        RadioButton cb = (RadioButton) layoutInflater.inflate(R.layout.checkbox, null);
        cb.setText(text);
        cb.setHeight(this.getResources().getDimensionPixelSize(R.dimen.cb_height));
        container.addView(cb);
        return cb;
    }

    private void addTitleView(ViewGroup container, String title) {
        layoutInflater.inflate(R.layout.container_title, container);
        TextView titleView = container.findViewById(R.id.title);
        titleView.setText(title);
    }

    private void setListeners() {
        Disposable disposable = RxView.clicks(playButton)
                .map(unit -> stateSubject.blockingFirst())
                .subscribe(state -> {
                    switch (state) {
                        case Idle:
                        case Recording: {
                            actionPublishSubject.onNext(new PlayAction());
                            break;
                        }
                        case Playing: {
                            actionPublishSubject.onNext(new StopAction());
                            break;
                        }
                    }
                });

        Disposable disposable1 = RxView.clicks(recordButton)
                .map(unit -> stateSubject.blockingFirst())
                .subscribe(state -> {
                    switch (state) {
                        case Idle:
                        case Playing: {
                            actionPublishSubject.onNext(new RecordAction());
                            break;
                        }
                        case Recording: {
                            actionPublishSubject.onNext(new StopAction());
                            break;
                        }
                    }
                });

        Disposable disposable2 = RxView.clicks(shareButton)
                .subscribe(unit ->
                        actionPublishSubject.onNext(new ShareAction())
                );

        compositeDisposable.addAll(disposable, disposable1, disposable2);

        for (int i = 0; i < bitRateGroup.size(); i++) {
            final int index = i;
            RadioButton radioButton = bitRateGroup.get(index);
            Disposable disposableRadio = RxView.clicks(radioButton)
                    .subscribe(unit -> {
                        VoiceRecorder.BitRate bitRate = VoiceRecorder
                                .BitRate.values()[index];
                        actionPublishSubject.onNext(new BitRateChangeAction(bitRate));
                    });
            compositeDisposable.add(disposableRadio);
        }

        for (int i = 0; i < sampleRateGroup.size(); i++) {
            final int index = i;
            RadioButton radioButton = sampleRateGroup.get(index);
            Disposable disposableRadio = RxView.clicks(radioButton)
                    .subscribe(unit -> {
                        VoiceRecorder.SampleRate sampleRate = VoiceRecorder
                                .SampleRate.values()[index];
                        actionPublishSubject.onNext(new SampleRateChangeAction(sampleRate));

                    });
            compositeDisposable.add(disposableRadio);
        }
    }

    public void renderView(MainViewModel mainViewModel) {
        renderPermissions(mainViewModel.requestForPermissions);
        renderBitrate(mainViewModel.bitRate);
        renderSampleRate(mainViewModel.sampleRate);
        renderError(mainViewModel.error);
        renderMessage(mainViewModel.message);
        renderState(mainViewModel.state);
        renderPlayerId(mainViewModel.playerId);
    }

    private void renderPlayerId(@Nullable Integer playerId) {
        if (playerId == null) {
            return;
        }
        visualizer = new Visualizer(playerId);
        visualizer.setCaptureSize(Visualizer.getCaptureSizeRange()[1]);
        visualizer.setDataCaptureListener(new Visualizer.OnDataCaptureListener() {
            public void onWaveFormDataCapture(Visualizer visualizer, byte[] bytes, int samplingRate) {
                visualizerView.updateVisualizer(bytes);
            }

            public void onFftDataCapture(Visualizer visualizer, byte[] bytes, int samplingRate) {
            }
        }, Visualizer.getMaxCaptureRate() / 2, true, false);
        visualizer.setEnabled(true);
    }

    private void stopVisualizer() {
        Disposable disposable = Completable
                .fromAction(() -> {
                    if (visualizer != null) {
                        visualizer.release();
                        visualizer = null;
                    }
                })
                .subscribeOn(computation())
                .subscribe();
        compositeDisposable.add(disposable);
    }

    private void renderState(State state) {
        stateSubject.onNext(state);
        playButton.setImageResource(state == State.Playing ?
                R.drawable.ic_action_stop :
                R.drawable.ic_action_play);

        recordButton.setImageResource(state == State.Recording ?
                R.drawable.ic_action_stop :
                R.drawable.ic_action_rec);
        if (state == State.Idle) stopVisualizer();
    }

    private void renderMessage(@Nullable String message) {
        if (message == null) {
            return;
        }
        @SuppressLint("InflateParams")
        TextView inflate = (TextView) getLayoutInflater().inflate(
                R.layout.log_view,
                null
        );
        logContainerView.addView(inflate);
        inflate.setText(message);
        scrollDownHandler.postDelayed(() -> scrollView.fullScroll(View.FOCUS_DOWN), DELAY_MILLIS);
    }

    private void renderError(@Nullable String error) {
        if (error == null) {
            return;
        }
        @SuppressLint("InflateParams")
        TextView inflate = (TextView) getLayoutInflater().inflate(R.layout.log_view, null);
        logContainerView.addView(inflate);
        inflate.setText(getString(
                R.string.error_string,
                error
        ));
        inflate.setTextColor(ContextCompat.getColor(this,
                android.R.color.holo_orange_light
        ));
        scrollDownHandler.postDelayed(() ->
                        scrollView.fullScroll(View.FOCUS_DOWN), DELAY_MILLIS
        );
    }

    private void renderPermissions(@Nullable Set<String> requestForPermissions) {
        if (requestForPermissions == null) {
            return;
        }
        HashMap<String, PermissionOptionalDetails> permissionRequests = new HashMap<>();
        for (String permissionName : requestForPermissions) {
            permissionRequests.put(permissionName, permissionsMap.get(permissionName));
        }
        permissionsManager.addPermissions(permissionRequests);
        permissionsManager.makePermissionRequest(true);
    }

    private void renderSampleRate(VoiceRecorder.SampleRate sampleRate) {
        for (int i = 0; i < sampleRateGroup.size(); i++) {
            sampleRateGroup.get(i).setChecked(sampleRate.ordinal() == i);
        }
    }

    private void renderBitrate(VoiceRecorder.BitRate bitRate) {
        for (int i = 0; i < bitRateGroup.size(); i++) {
            bitRateGroup.get(i).setChecked(bitRate.ordinal() == i);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        compositeDisposable.clear();
        scrollDownHandler.removeCallbacksAndMessages(null);
        stopVisualizer();
    }
}

File: app/src/main/java/com/omar/retromp3recorder/app/main/MainActivity.kt
package com.omar.retromp3recorder.app.main

import android.Manifest
import android.annotation.SuppressLint
import android.media.audiofx.Visualizer
import android.media.audiofx.Visualizer.OnDataCaptureListener
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.github.alkurop.jpermissionmanager.PermissionOptionalDetails
import com.github.alkurop.jpermissionmanager.PermissionRequiredDetails
import com.github.alkurop.jpermissionmanager.PermissionsManager
import com.jakewharton.rxbinding3.view.clicks
import com.omar.retromp3recorder.app.App
import com.omar.retromp3recorder.app.R
import com.omar.retromp3recorder.app.customviews.VisualizerView
import com.omar.retromp3recorder.app.recording.recorder.Mp3VoiceRecorder
import com.omar.retromp3recorder.app.recording.recorder.Mp3VoiceRecorder.SampleRate
import com.squareup.picasso.Picasso
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject
import java.util.*
import javax.inject.Inject

class MainActivity : AppCompatActivity(R.layout.activity_main) {
    private val actionPublishSubject = PublishSubject.create<MainView.Action>()
    private val compositeDisposable = CompositeDisposable()
    private val scrollDownHandler = Handler(Looper.getMainLooper())
    private val stateSubject: Subject<MainView.State> = BehaviorSubject.create()

    private val playButton: ImageView by lazy { findViewById(R.id.iv_play) }
    private val recordButton: ImageView by lazy { findViewById(R.id.iv_record) }
    private val shareButton: ImageView by lazy { findViewById(R.id.iv_share) }

    private val logContainerView: LinearLayout by lazy { findViewById(R.id.ll_logHolder) }
    private val sampleRateContainer: LinearLayout by lazy { findViewById(R.id.left_radio_container) }
    private val bitRateContainer: LinearLayout by lazy { findViewById(R.id.right_radio_container) }

    private val scrollView: ScrollView by lazy { findViewById(R.id.scrollView) }
    private val visualizerView: VisualizerView by lazy { findViewById(R.id.visualizer) }
    private val background: ImageView by lazy { findViewById(R.id.background) }

    private val sampleRateGroup: List<RadioButton> by lazy { createSampleRateGroup() }
    private val bitRateGroup: List<RadioButton> by lazy { createBitRateGroup() }

    private val checkboxHeight by lazy { resources.getDimensionPixelSize(R.dimen.cb_height) }

    private val permissionsManager: PermissionsManager by lazy { PermissionsManager(this) }
    private val permissionsMap: Map<String, PermissionOptionalDetails> by lazy { createPermissionsMap() }

    private var visualizer: Visualizer? = null

    @Inject
    lateinit var interactor: MainViewInteractor

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        App.appComponent.inject(this)
        addTitleView(bitRateContainer, getString(R.string.bit_rate))
        addTitleView(sampleRateContainer, getString(R.string.sample_rate))
        setUpButtonListeners()
        Picasso.with(this).load(R.drawable.bg).fit().into(background)
        val disposable = actionPublishSubject
            .compose(interactor.process())
            .compose(MainViewResultMapper.map())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { mainViewModel: MainView.MainViewModel ->
                renderView(
                    mainViewModel
                )
            }
        compositeDisposable.add(disposable)
    }

    private fun renderView(mainViewModel: MainView.MainViewModel) {
        renderPermissions(mainViewModel.requestForPermissions.ghost)
        renderBitrate(mainViewModel.bitRate)
        renderSampleRate(mainViewModel.sampleRate)
        renderError(mainViewModel.error.ghost?.bell(this))
        renderMessage(mainViewModel.message.ghost?.bell(this))
        renderState(mainViewModel.state)
        renderPlayerId(mainViewModel.playerId.ghost)
    }

    private fun addTitleView(container: ViewGroup, title: String) {
        layoutInflater.inflate(R.layout.container_title, container)
        val titleView = container.findViewById<TextView>(R.id.title)
        titleView.text = title
    }

    private fun createSampleRateGroup(): List<RadioButton> =
        SampleRate.values().map { sampleRate ->
            sampleRateContainer.addCheckBox(
                title = getString(
                    R.string.sample_rate_format,
                    sampleRate.value
                )
            )
        }

    private fun createBitRateGroup(): List<RadioButton> =
        Mp3VoiceRecorder.BitRate.values().map { bitRate ->
            bitRateContainer.addCheckBox(
                title = getString(
                    R.string.bit_rate_format,
                    bitRate.value
                )
            )
        }

    @SuppressLint("InflateParams")
    private fun ViewGroup.addCheckBox(title: String): RadioButton {
        val container = this
        return (layoutInflater.inflate(R.layout.checkbox, null) as RadioButton).apply {
            this.text = title
            this.height = checkboxHeight
            container.addView(this)
        }
    }

    private fun createPermissionsMap(): Map<String, PermissionOptionalDetails> =
        listOf(
            Pair(
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                PermissionRequiredDetails(
                    getString(R.string.write_permission_title),
                    getString(R.string.write_permission_message),
                    getString(R.string.write_required_message)
                )
            ),
            Pair(
                Manifest.permission.RECORD_AUDIO,
                PermissionRequiredDetails(
                    getString(R.string.record_permission_title),
                    getString(R.string.record_permission_message),
                    getString(R.string.record_required_message)
                )
            )
        ).toMap()

    private fun setUpButtonListeners() {
        compositeDisposable.addAll(
            playButton.clicks().map { stateSubject.blockingFirst() }
                .subscribe { state ->
                    when (state) {
                        MainView.State.Idle,
                        MainView.State.Recording -> actionPublishSubject.onNext(MainView.Action.Play)
                        MainView.State.Playing -> actionPublishSubject.onNext(MainView.Action.Stop)
                    }
                },
            recordButton.clicks().map { stateSubject.blockingFirst() }
                .subscribe { state ->
                    when (state) {
                        MainView.State.Idle,
                        MainView.State.Playing -> actionPublishSubject.onNext(MainView.Action.Record)
                        MainView.State.Recording -> actionPublishSubject.onNext(MainView.Action.Stop)
                    }
                },
            shareButton.clicks().subscribe { actionPublishSubject.onNext(MainView.Action.Share) }
        )
        bitRateGroup.mapIndexed { index, radioButton ->
            radioButton.clicks().subscribe {
                val bitRate = Mp3VoiceRecorder.BitRate.values()[index]
                actionPublishSubject.onNext(MainView.Action.BitRateChange(bitRate))
            }
        }.forEach { compositeDisposable.add(it) }

        sampleRateGroup.mapIndexed { index, radioButton ->
            radioButton.clicks()
                .subscribe {
                    val sampleRate = SampleRate.values()[index]
                    actionPublishSubject.onNext(MainView.Action.SampleRateChange(sampleRate))
                }
        }.forEach { compositeDisposable.add(it) }
    }

    private fun renderPlayerId(playerId: Int?) {
        if (playerId == null) {
            return
        }
        visualizer = Visualizer(playerId).apply {
            captureSize = Visualizer.getCaptureSizeRange()[1]
            setDataCaptureListener(object : OnDataCaptureListener {
                override fun onWaveFormDataCapture(
                    visualizer: Visualizer,
                    bytes: ByteArray,
                    samplingRate: Int
                ) = visualizerView.updateVisualizer(bytes)

                override fun onFftDataCapture(
                    visualizer: Visualizer,
                    bytes: ByteArray,
                    samplingRate: Int
                ) = Unit
            }, Visualizer.getMaxCaptureRate() / 2, true, false)
            enabled = true
        }
    }

    private fun stopVisualizer() {
        compositeDisposable.add(Completable
            .fromAction {
                if (visualizer != null) {
                    visualizer?.release()
                    visualizer = null
                }
            }
            .subscribeOn(Schedulers.computation())
            .subscribe())
    }

    private fun renderState(state: MainView.State) {
        stateSubject.onNext(state)
        playButton.setImageResource(
            if (state == MainView.State.Playing) R.drawable.ic_action_stop
            else R.drawable.ic_action_play
        )
        recordButton.setImageResource(
            if (state == MainView.State.Recording) R.drawable.ic_action_stop
            else R.drawable.ic_action_rec
        )
        if (state == MainView.State.Idle) stopVisualizer()
    }

    private fun renderMessage(message: String?) {
        if (message == null) {
            return
        }
        @SuppressLint("InflateParams") val inflate = layoutInflater.inflate(
            R.layout.log_view,
            null
        ) as TextView
        logContainerView.addView(inflate)
        inflate.text = message
        scrollDownHandler.postDelayed(
            { scrollView.fullScroll(View.FOCUS_DOWN) },
            DELAY_MILLIS.toLong()
        )
    }

    private fun renderError(error: String?) {
        if (error == null) {
            return
        }
        @SuppressLint("InflateParams") val inflate =
            layoutInflater.inflate(R.layout.log_view, null) as TextView
        logContainerView.addView(inflate)
        inflate.text = getString(
            R.string.error_string,
            error
        )
        inflate.setTextColor(
            ContextCompat.getColor(
                this,
                android.R.color.holo_orange_light
            )
        )
        scrollDownHandler.postDelayed(
            { scrollView.fullScroll(View.FOCUS_DOWN) }, DELAY_MILLIS
                .toLong()
        )
    }

    private fun renderPermissions(requestForPermissions: Set<String>?) {
        if (requestForPermissions == null) {
            return
        }
        val permissionRequests = HashMap<String, PermissionOptionalDetails?>()
        for (permissionName in requestForPermissions) {
            permissionRequests[permissionName] = permissionsMap[permissionName]
        }
        permissionsManager.addPermissions(permissionRequests)
        permissionsManager.makePermissionRequest(true)
    }

    private fun renderSampleRate(sampleRate: SampleRate) {
        sampleRateGroup.forEachIndexed { index, radioButton ->
            radioButton.isChecked = sampleRate.ordinal == index
        }
    }

    private fun renderBitrate(bitRate: Mp3VoiceRecorder.BitRate) {
        bitRateGroup.forEachIndexed { index, radioButton ->
            radioButton.isChecked = bitRate.ordinal == index
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
        scrollDownHandler.removeCallbacksAndMessages(null)
        stopVisualizer()
    }
}

private const val DELAY_MILLIS = 150
